lib/icingadb/icingadb.cpp

```diff
@@ -52,9 +57,45 @@ void IcingaDB::Start(bool runtimeCreated)
 {
 	ObjectImpl<IcingaDB>::Start(runtimeCreated);
 
-	boost::call_once([]() {
-		m_EnvironmentId = SHA1(GetEnvironment());
-	}, m_EnvironmentIdOnce);
+	std::call_once(m_EnvironmentIdOnce, []() {
+		String path = Configuration::DataDir + "/icingadb.env";
+
+		if (Utility::PathExists(path)) {
+			std::ifstream fp(path);
+			fp >> m_EnvironmentId;
+
+			if (m_EnvironmentId.GetLength() != 2*SHA_DIGEST_LENGTH) {
+				throw std::runtime_error("Wrong length of stored Icinga DB environment");
+			}
+
+			for (unsigned char c : m_EnvironmentId) {
+				if (!std::isxdigit(c)) {
+					throw std::runtime_error("Stored Icinga DB environment is not a hex string");
+				}
+			}
+		} else {
+			std::shared_ptr<X509> cert = GetX509Certificate(ApiListener::GetDefaultCaPath());
+
+			unsigned int n;
+			unsigned char digest[EVP_MAX_MD_SIZE];
+			if (X509_pubkey_digest(cert.get(), EVP_sha1(), digest, &n) != 1) {
+				BOOST_THROW_EXCEPTION(openssl_error()
+					<< boost::errinfo_api_function("X509_pubkey_digest")
+					<< errinfo_openssl_error(ERR_peek_error()));
+			}
+
+			m_EnvironmentId = BinaryToHex(digest, n);
+
+			std::fstream fp;
+			String tmp = Utility::CreateTempFile(path + ".XXXXXX", 0600, fp);
+			fp << m_EnvironmentId << std::endl;
+			fp.close();
```

Comment:
And fsync(2)?

Reply:
Which we do nowhere in the code so far. Sure, that doesn't mean that we shouldn't do so (looking at #8840, yes, that's Windows, but should maybe also be extended to Linux and we might just have been lucky there so far). For now I'd go with [`{Load,Save}JsonFile`](https://github.com/Icinga/icinga2/pull/9036#discussion_r743595751) and move that question to these functions.